### PR TITLE
ISPN-14670 HRANDFIELD + ISPN-14668 HMGET

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-14670
https://issues.redhat.com/browse/ISPN-14668

### CODE REVIEW

File: multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapPairCache.java [hunk 1]
```diff
@@ -194,4 +226,45 @@ public final CompletionStage<Integer> remove(K key, Collection<HK> properties) {
...
+    * @return {@link CompletionStage} containing a {@link Map} with the key-value pairs or null if not found.
+    */
+   public CompletionStage<Map<HK, HV>> subSelect(K key, int count) {
+      requireNonNull(key, ERR_KEY_CAN_T_BE_NULL);
+      requirePositive(count, ERR_COUNT_MUST_BE_POSITIVE);
+      return cache.getCacheEntryAsync(key)
+            .thenApply(entry -> {
+               if (entry == null) return null;
+
+               return entry.getValue().converted().entrySet().stream()
```
Comment (wburns) [inline multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapPairCache.java:247]: if count is equal to or greater can we just return the map?

File: multimap/src/test/java/org/infinispan/multimap/impl/DistributedMultimapPairCacheTest.java [hunk 2]
```diff
@@ -138,4 +145,22 @@ public void testContainsProperty() {
...
+   public void testSubSelect() {
+      assertFromAllCaches(multimap -> await(multimap.set("sub-select-test", Map.entry(toBytes("oihana"), OIHANA), Map.entry(toBytes("koldo"), KOLDO))
+            .thenCompose(ignore -> multimap.subSelect("sub-select-test", 1))
+            .thenAccept(m ->
+                  assertThat(convertKeys(m))
+                        .hasSize(1)
+                        .containsAnyOf(Map.entry("oihana", OIHANA), Map.entry("koldo", KOLDO)))));
+   }
+
+   public void testGetMultiple() {
```
Comment (wburns) [inline multimap/src/test/java/org/infinispan/multimap/impl/DistributedMultimapPairCacheTest.java:158]: Should we add a test for what it returns when the key isn't present? Should we return an entry with a null key or just not return it?

File: server/resp/src/main/java/org/infinispan/server/resp/ByteBufferUtils.java [hunk 3]
```diff
@@ -75,10 +75,16 @@ public static ByteBuf writeLong(Long result, ByteBufPool alloc) {
...
+      if (results.isEmpty())
+         return stringToByteBuf("*0\r\n", alloc);
+
       int resultBytesSize = 0;
       for (byte[] result: results) {
-         int length = result.length;
-         if (length > 0) {
+         int length;
+         if (result == null) {
+            // $-1
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/ByteBufferUtils.java:85]: :+1: Looks like a different bug

File: server/resp/src/main/java/org/infinispan/server/resp/commands/hash/HMGET.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+   public HMGET() {
+      super(-3, 1, 1, 1);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx, List<byte[]> arguments) {
+      EmbeddedMultimapPairCache<byte[], byte[], byte[]> multimap = handler.getHashMapMultimap();
+      CompletionStage<Collection<byte[]>> cs = multimap.get(arguments.get(0), arguments.subList(1, arguments.size()).toArray(new byte[0][]))
+            .thenApply(entries -> {
+               List<byte[]> result = new ArrayList<>(arguments.size());
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/hash/HMGET.java:39]:
```suggestion
               List<byte[]> result = new ArrayList<>(arguments.size() - 1);
```

File: server/resp/src/main/java/org/infinispan/server/resp/commands/hash/HRANDFIELD.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            RespErrorUtil.syntaxError(handler.allocator());
+            return handler.myStage();
+         }
+         withValues = true;
+      }
+
+      EmbeddedMultimapPairCache<byte[], byte[], byte[]> multimap = handler.getHashMapMultimap();
+      int finalCount = count;
+      boolean finalWithValues = withValues;
+      CompletionStage<Collection<byte[]>> cs = multimap.subSelect(key, Math.abs(count))
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/hash/HRANDFIELD.java:70]: One thing I worry is that repeated invocations of a count < map size will always return the same subset of keys just in random orders. I don't think that is quite correct though.

**Review by wburns — CHANGES_REQUESTED** [review]

File: multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapPairCache.java (hunk 1 above)
Comment (jabolina) [reply multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapPairCache.java]: Yep, updated.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/hash/HRANDFIELD.java (hunk 5 above)
Comment (jabolina) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/hash/HRANDFIELD.java:70]: Updated to wrap the entry Set in a List and use the Collections.shuffle.

File: server/resp/src/main/java/org/infinispan/server/resp/ByteBufferUtils.java (hunk 3 above)
Comment (jabolina) [reply server/resp/src/main/java/org/infinispan/server/resp/ByteBufferUtils.java:85]: I believe the method assumed non-null elements. Should I log a JIRA?

Comment (wburns) [reply server/resp/src/main/java/org/infinispan/server/resp/ByteBufferUtils.java:85]: I guess it is fine. Nothing from the method made me assume it only allowed non null values though. Looks like it would just throw a NPE and we had to rely upon tests to catch it :) Either way no reason for this method not to support a null value.

**Comment by wburns** [comment]
Integrated into main, thanks @jabolina !
